---
title: An Introduction to the GenomicAlignments Package
author: Hervé Pagès
clean: false
output:
  html_document:
    toc: yes
    df_print: paged
    number_sections: yes
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

# Introduction

The `GenomicAlignments` package serves as the foundation for representing genomic alignments within the **Bioconductor** project.  In the **Bioconductor** package hierarchy, it builds upon the `GenomicRanges` (infrastructure) package and provides support for many **Bioconductor** packages. This package defines three classes: `GAlignments`, `GAlignmentPairs`, and `GAlignmentsList`), which are used to represent genomic alignments, pairs of genomic alignments, and groups of genomic alignments.

The `GenomicAlignments` package is available at bioconductor.org and can be downloaded via *BiocManager::install*:

```{r install, eval=FALSE}
if (!require("BiocManager"))
    install.packages("BiocManager")
BiocManager::install("GenomicAlignments")
```

```{r initialize, results='hide', message=FALSE, warning=FALSE}
library(GenomicAlignments)
```


# `GAlignments`: Genomic Alignments

The `GAlignments` class which is a container for storing a set of genomic alignments. The class is intended to support alignments in general, not only those coming from a 'Binary Alignment Map' or 'BAM' files. Also alignments with gaps in the reference sequence (a.k.a. **gapped alignments**) are supported which, for example, makes the class suited for storing junction reads from an RNA-seq experiment. More precisely, a `GAlignments` object is a vector-like object where each element describes an **alignment**, that is,
how a given sequence (called **query** or **read**, typically short) aligns to a reference sequence (typically long). As shown later in this document, a `GAlignments` object can be created from a 'BAM' file. In that case, each element in the resulting object will correspond to a record in the file. 

One important thing to note though is that not all the information present in the BAM/SAM records is stored in the object. In particular, for now, we discard the query sequences (SEQ field), the query ids (QNAME field), the query qualities (QUAL), the mapping qualities (MAPQ) and any other information that is not needed in order to support the basic set of operations described in this document.

This also means that multi-reads (i.e. reads with multiple hits in the reference) don't receive any special treatment i.e. the various SAM/BAM records corresponding to a multi-read will show up in the `GAlignments` object as if they were coming from different/unrelated queries. Also paired-end reads will be treated as single-end reads and the pairing information will be lost. This might change in the future.

## Load a `BAM' file into a `GAlignments` object

First we use the `readGAlignments` function from the `GenomicAlignments` package to load a toy `BAM' file into a `GAlignments` object:

```{r readGAlignments, message=FALSE, warning=FALSE}
library(GenomicAlignments)
aln1_file <- system.file("extdata", "ex1.bam", package="Rsamtools")
aln1 <- readGAlignments(aln1_file)
aln1
length(aln1)
```

`r length(aln1)[1]` 'BAM' records were loaded into the object. Note that `readGAlignments` would have discarded any `BAM' record describing an unaligned query (see description of the <flag> field in the SAM Format Specification ^[http://samtools.sourceforge.net/SAM1.pdf] for more information).

The reader interested in tracking down these events can always use the `scanBam` function but this goes beyond the scope of this document.

## Simple accessor methods

There is one accessor per field displayed by the `show` method and it has the same name as the field. All of them return a vector or factor of the same length as the object:

```{r accessors}
head(seqnames(aln1))
seqlevels(aln1)
head(strand(aln1))
head(cigar(aln1))
head(qwidth(aln1))
head(start(aln1))
head(end(aln1))
head(width(aln1))
head(njunc(aln1))
```

## More accessor methods

[coming soon...]

# `GAlignmentPairs`: Pairs of Genomic Alignments

[coming soon...]

# `GAlignmentsList`: Groups of Genomic Alignments

[coming soon...]

# Session Information

All of the output in this vignette was produced under the following conditions:

```{r SessionInfo}
sessionInfo()
```

